src/puya/ir/optimize/compiled_reference.py

```diff
@@ -0,0 +1,163 @@
+            except KeyError:
+                raise InternalError(
+                    f"Invalid contract reference: {const.artifact}", const.source_location
+                ) from None
+            match field:
+                case TxnField.GlobalNumUint:
+                    total = state_total.global_uints
+                case TxnField.GlobalNumByteSlice:
+                    total = state_total.global_bytes
+                case TxnField.LocalNumUint:
+                    total = state_total.local_uints
+                case TxnField.LocalNumByteSlice:
+                    total = state_total.local_bytes
+                case _:
+                    raise InternalError(
+                        f"Invalid state total field: {field.name}", const.source_location
+                    )
+            return ir.UInt64Constant(
+                value=total,
+                source_location=const.source_location,
+            )
+
+        if not _is_constant(const.template_variables):
+            return const
+        template_constants = _get_template_constants(
+            self.context.options.template_variables, const.template_variables
+        )
+        match field:
+            case TxnField.ApprovalProgramPages | TxnField.ClearStateProgramPages:
+                page = const.program_page
+                if page is None:
+                    raise InternalError("expected non-none value for page", const.source_location)
+                program_bytecode = self.context.get_program_bytecode(
+                    const.artifact,
+                    ("approval" if field == TxnField.ApprovalProgramPages else "clear_state"),
+                    template_constants,
+                )
+                program_page = program_bytecode[
+                    page * MAX_BYTES_LENGTH : (page + 1) * MAX_BYTES_LENGTH
+                ]
+                return ir.BytesConstant(
+                    value=program_page,
+                    encoding=AVMBytesEncoding.base64,
+                    source_location=const.source_location,
+                )
+            case TxnField.ExtraProgramPages:
+                approval_bytecode = self.context.get_program_bytecode(
+                    const.artifact, "approval", template_constants
+                )
+                clear_bytecode = self.context.get_program_bytecode(
+                    const.artifact, "clear_state", template_constants
+                )
+                total_bytes = len(approval_bytecode) + len(clear_bytecode)
+                extra_pages = (total_bytes - 1) // MAX_APP_PAGE_SIZE
+                return ir.UInt64Constant(
+                    value=extra_pages,
+                    source_location=const.source_location,
+                )
+        raise InternalError(
+            f"Unhandled compiled reference field: {field.name}", const.source_location
+        )
+
+
+def _is_constant(template_variables: Mapping[str, ir.Value]) -> bool:
```

Comment:
can we make this a `typing.TypeGuard` return? then we can use `typing.assert_never` on line 160 below, instead of an `InternalError`?